baremetal/metal3machine_manager.go

```diff
@@ -1338,15 +1335,58 @@ func (m *MachineManager) SetNodeProviderID(ctx context.Context, bmhID, providerI
 		m.Log.Info("Target node is not found, requeuing")
 		return &RequeueAfterError{RequeueAfter: requeueAfter}
 	}
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+	bmhName, err := m.getBmhNameFromM3Machine()
+	if err != nil {
+		m.Log.Info("unable to retrieve BMH from Metal3Cluster")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	providerIDLegacy := *providerID
+	providerIDNew := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	if !m.Metal3Cluster.Spec.NoCloudProvider {
+		// Settig providerID on nodes takes time. Continuous monitoring is required.
+		nodesWithProviderIDset := 0
+		for _, node := range nodes.Items {
+			m.Log.Info("PP0: ", node.Spec.ProviderID)
+			if node.Spec.ProviderID == "" {
+				m.Log.Info("PP1: NoCloudProvider set to false. But, no providerID is set on the node either", node.GetName())
+				nodesWithProviderIDset++
+			} else if node.Spec.ProviderID == providerIDNew {
+				m.Log.Info("PP2: Found a node with matching node providerID in new format")
+				*providerID = providerIDNew
+				return nil
+			} else if node.Spec.ProviderID == providerIDLegacy {
+				m.Log.Info("PP3: Found a node with matching node providerID in legacy format")
+				*providerID = providerIDLegacy
+				return nil
+			}
+		}
+		// We have considered all nodes, and no matching node was found.
+		if nodesWithProviderIDset == nodesCount {
+			return errors.New("No matching node with the providerID found: ")
+		}
+		m.Log.Info("Waiting for more nodes to have their providerID property set before declaring failure")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	var nodeVar corev1.Node
 	for _, node := range nodes.Items {
-		node := node
-		if node.Spec.ProviderID == providerID {
-			continue
+		if node.Spec.ProviderID == "" {
+			m.Log.Info("missing node providerID, setting value from label")
+			node.Spec.ProviderID = providerIDLegacy
```

Comment:
here we could set the new format, no ?

Reply:
I was not sure on that, but I assumed this:
There are two sources or providerID (node level) and (code level). If the former is missing, the latter is used. 
But, I can change it to the new format as well.

Reply:
here, the providerID is unset, and you identified the node based on the label. It is really up to you what providerID you set there (as long as it is the same as on the m3m) . Since we introduce a new format, let's use it by default.